modules/prebidServerBidAdapter.js

```diff
@@ -272,126 +261,289 @@ export function PrebidServer() {
+            bidObject.dealId = bidObj.deal_id;
+          }
+          bidObject.requestId = bidObj.bid_id;
+          bidObject.creativeId = bidObj.creative_id;
+
+          // TODO: Remove when prebid-server returns ttl, currency and netRevenue
+          bidObject.ttl = (bidObj.ttl) ? bidObj.ttl : DEFAULT_S2S_TTL;
+          bidObject.currency = (bidObj.currency) ? bidObj.currency : DEFAULT_S2S_CURRENCY;
+          bidObject.netRevenue = (bidObj.netRevenue) ? bidObj.netRevenue : DEFAULT_S2S_NETREVENUE;
+
+          bids.push({ adUnit: bidObj.code, bid: bidObject });
+        });
+      }
     }
-    const payload = JSON.stringify(requestJson);
-    ajax(_s2sConfig.endpoint, processResponse, payload, {
-      contentType: 'text/plain',
-      withCredentials: true
-    });
-  };
 
-  // at this point ad units should have a size array either directly or mapped so filter for that
-  function hasSizes(unit) {
-    return unit.sizes && unit.sizes.length;
+    return bids;
   }
+};
 
-  /* Notify Prebid of bid responses so bids can get in the auction */
-  function handleResponse(response, requestedBidders, bidRequests, addBidResponse, done) {
-    let result;
-    try {
-      result = JSON.parse(response);
+/*
+ * Protocol spec for OpenRTB endpoint
+ * e.g., https://<prebid-server-url>/v1/openrtb2/auction
+ */
+const OPEN_RTB_PROTOCOL = {
 
-      if (result.status === 'OK' || result.status === 'no_cookie') {
-        if (result.bidder_status) {
-          result.bidder_status.forEach(bidder => {
-            if (bidder.no_cookie) {
-              doBidderSync(bidder.usersync.type, bidder.usersync.url, bidder.bidder);
-            }
-          });
-        }
+  bidMap: {},
 
-        // do client-side syncs if available
-        requestedBidders.forEach(bidder => {
-          let clientAdapter = adaptermanager.getBidAdapter(bidder);
-          if (clientAdapter && clientAdapter.registerSyncs) {
-            clientAdapter.registerSyncs([]);
+  buildRequest(s2sBidRequest, adUnits) {
+    let imps = [];
+
+    // transform ad unit into array of OpenRTB impression objects
+    adUnits.forEach(adUnit => {
+      let banner;
+
+      adUnit.bids.forEach(bid => {
+        // OpenRTB response contains the adunit code and bidder name. These are
+        // combined to create a unique key for each bid since an id isn't returned
+        const key = `${adUnit.code}${bid.bidder}`;
+        this.bidMap[key] = bid;
+      });
+
+      const bannerParams = utils.deepAccess(adUnit, 'mediaTypes.banner');
+      if (bannerParams && bannerParams.sizes) {
+        // get banner sizes in form [{ w: <int>, h: <int> }, ...]
+        const format = bannerParams.sizes.reduce((acc, size) =>
+          [...acc, { w: size[0], h: size[1] }], []);
+
+        banner = {format};
+      }
+
+      // get bidder params in form { <bidder code>: {...params} }
+      const ext = adUnit.bids.reduce((acc, bid) =>
+        Object.assign({}, acc, { [bid.bidder]: bid.params }), {});
```

Comment:
Same performance complaint as a above.

```javascript
const ext = adUnit.bids.reduce((acc, bid) => {
	acc[bid.bidder] = bid.params;
	return acc;
}, {})
```
In my opinion this is more readable, but opinion aside it's _definitely_ going to perform much better as it's not doing additional iteration per each loop (each arg passed to `Object.assign` is extra iteration) and not instantiating a bunch of throwaway object literals.

Reply:
Ah yeah, totally blanked on the way of doing this, thanks, updated